Mrc 6821 serve region metadata

EmmaLRussell opened the pull request:
**No hurry on reviewing this - I don't expect to be back on grout before the new year...**

This branch adds endpoints to serve region metadata. Metadata is embedded in the tiles, but not easily accessible so it's useful to have the region metadata explicitly provided through json endpoints. Metadata consists of id, name, bounds and any additional feature properties from the original geojson. 

The region metadata itself is generated in a new report in `grout-orderly` (in t[his PR](https://github.com/mrc-ide/grout-orderly/pull/4)) - I've included partial output from this in the `region-metadata` folder for local dev testsing. Each call to the endpoint simply reads and serves the contents of the relevant file (we might do a bit of caching later!). 

There are two endpoints:
* `/region-metadata/:dataset/admin0` provides country level (admin 0) metadata for all countries. 
* `/region-metadata/:dataset/:level/:iso` provides metadata for all the regions of level `:level` within the country represented by `:iso`

I've also included discovery of the relevant region metadata folders when building the app metadata (*sigh* unfortunately we now have two levels of metadata...) which is returned by the `/metadata` endpoint.

EmmaLRussell committed:
added region metadata controller and routes

EmmaLRussell committed:
error handling

EmmaLRussell committed:
region metadata files

EmmaLRussell committed:
filename updates

EmmaLRussell committed:
integration test

EmmaLRussell committed:
add unit test

EmmaLRussell committed:
lint

EmmaLRussell committed:
update README

EmmaLRussell committed:
include region-metadata when run in docker

EmmaLRussell committed:
rename level folders to be consistent with tiles

EmmaLRussell committed:
add region metadata datasets to /metadata endpoint

EmmaLRussell committed:
add discoverRegionMetadata unit test

EmmaLRussell committed:
lint

david-mears-2 commented on schema/Metadata.schema.json:
```diff
@@ -5,24 +5,28 @@
-            "required": ["levels"]
-          }
-        }
+        "tile": { "$ref": "#/$defs/datasets" },
+        "regionMetadata": { "$ref": "#/$defs/datasets" }
```
I don't understand the difference between 'tile' and 'regionMetadata'. In the endpoint, these return the same thing. I infer from the fact that both exist, that they can in theory be different: is that true?

david-mears-2 replied:
Also, all of the data is sourced from a root-level folder called `region_metadata`, which, I surmise, doesn't correspond to the `regionMetadata` property because that's not the only property under 'datasets'. Should I expect to see a root-level folder called 'tile'?

Ah, I've just spotted there is a root-level folder called 'data' containing what looks like tile data.

I might have found it easier to follow if all data (included metadata) was stored under the root-level 'data' folder.

EmmaLRussell replied:
Yep, will do

david-mears-2 replied:
This feels easier to navigate now

david-mears-2 commented on src/controllers/regionMetadataController.ts:
```diff
@@ -0,0 +1,56 @@
+// We expect to find region metadata data in folders at ./region_metadata/:dataset/:level
+// All level 0 metadata should be concatenated in a single file named :dataset_0.json
+// Level 1 and level 2 metadata should be in one file per country iso named :dataset_:iso_:level.json
+export class RegionMetadataController {
```
Since 'country' and 'level0' mean the same thing, I think the naming of these functions is confusing. I would call the second one 'getLowerLevelMetadata'.

EmmaLRussell replied:
Ha, I find "lower level" more confusing because is it lower level as in more detailed or a lower admin level number..? How about something like `getGlobalLevel0Metadata` and `getMetadataForRequestISO`?

david-mears-2 replied:
Happy with that; why not tersify to `getISOMetadata` / `getMetadataByISO`?

david-mears-2 commented on README.md:
```diff
@@ -1,12 +1,24 @@
+* Metadata about available datasets: `/metadata`
+* Tile data: `/tile/:dataset/:level/:z/:x/:y"`
+* Admin 0 (country) metadata for all countries in a dataset: `/region-metadata/:dataset/admin0`
+* Region metadata for a single country: `/region-metadata/:dataset/:level/:iso`
```
(If not covered by making the /metadata response clearer) Can you explicitly document that providing admin0 for :level is not allowed?

david-mears-2 commented on tests/integration/metadata.spec.ts:
```diff
@@ -10,6 +10,11 @@ describe("metadata endpoint", () => {
+                regionMetadata: {
+                    gadm41: {
+                        levels: ["admin0", "admin1"]
+                    }
```
I think this test doesn't verify that the assignation of region metadata to `regionMetadata` and tile data to `tile` is happening, since the return value is the same.

david-mears-2 replied:
(Might not be worth bothering with, because integration test, so it's fiddly to create different mocks)

EmmaLRussell replied:
Yeah, I think that would be tricky because it's looking at the real responses - it should be covered in the unit test.

EmmaLRussell replied:
As I say, I'm going to revamp the metadata endpoint anyway.

david-mears-2 commented on tests/unit/controllers/regionMetadataController.spec.ts:
```diff
@@ -0,0 +1,86 @@
+    fs.writeFileSync(path, JSON.stringify(testMetadata));
+};
+
+describe("RegionMetadataController", () => {
```
I think this test suite doesn't verify that the two functions getLevel0Metadata and getCountryMetadata do different things, since the response value is the same. Can you make each test use a different `testMetadata`?

Related to earlier comment: I find the names of the two tests here just as ambiguous (because level 0 means country to me) as the function names.

david-mears-2 commented:
> (sigh unfortunately we now have two levels of metadata...)

Is the 'region metadata' actually metadata, or can we call it just data? Maybe it's meta from the perspective of 'force of infection data', but for grout, it could be considered non-meta? Not sure

--

If the /metadata endpoint tells me:

```json
{
  "status": "success",
  "errors": null,
  "data": {
    "datasets": {
      "tile": {
        "gadm41": {
          "levels": [
            "admin0",
            "admin1"
          ]
        }
      },
      "regionMetadata": {
        "gadm41": {
          "levels": [
            "admin0",
            "admin1"
          ]
        }
      }
    }
  }
}
```

Shouldn't I expect that a request to `/region-metadata/gadm41/admin0/ETH` is valid, a valid request for the admin0 region metadata for Ethiopia? That's how I'd naturally interpret the 'metadata' info: telling me what requests would be valid. If the region metadata endpoint is going to reject requests for level 0, I think the metadata endpoint should tell me:

```json
...
"regionMetadata": {
        "gadm41": {
          "levels": [
            "admin1"
          ]
        }
      }
...
```

EmmaLRussell commented:
> Is the 'region metadata' actually metadata, or can we call it just data? Maybe it's meta from the perspective of 'force of infection data', but for grout, it could be considered non-meta? Not sure

It's metadata in the sense that it's not geographical tile data (all the latlong points) which is the main data for grout.. But yeah, maybe it would make sense to include region metadata as a subset of data on disk. I think we still have "app metadata" and "region metadata" but hopefully the distinction is fairly clear..

EmmaLRussell commented:
> Shouldn't I expect that a request to /region-metadata/gadm41/admin0/ETH is valid, a valid request for the admin0 region metadata for Ethiopia? That's how I'd naturally interpret the 'metadata' info: telling me what requests would be valid. If the region metadata endpoint is going to reject requests for level 0, I think the metadata endpoint should tell me:

I wasn't really thinking about the metadata telling the client what endpoints are valid in itself, but giving some of the information they need  to work it out for themselves - they need the knowledge that admin0 is only provided by a global endpoint, and that they need to look at the response from that to find out what countries exist in the dataset, which they can then combine with the levels in the metadata response to work out what endpoints they can call for admin1 etc. 

But I think you're right that this is not very satisfactory as there is an inconsistency between the two types of response (data and region metadata) which we're glossing over by making them have the same metadata structure. I'm thinking maybe what we could do is actually be more explicit about the form of the valid requests and including that in the metadata - and also including relevant values e.g. all the valid `:iso` values. We can't do that for all possible endpoints e.g. the tile requests use obscure `:z/:x/:y` values, but we could certainly do it for `:iso`. I'll have a think about a better format..

EmmaLRussell commented:
For the metadata response, does something like this seem better? It's more explicit about the valid urls and how the metadata provided fits into them. It does need to cover the issue of region metadata having two types of url, a "global" one and a "country" one (or suggest better names for these!)

```
{
  "status": "success",
  "errors": null,
  "data": {
    "datasets": {
      "tile": {
        "gadm41": {
          "url: "/tile/gadm41/:level/:z/:x/:y",
          "level": [
            "admin0",
            "admin1"
          ]
        }
      },
      "regionMetadata": {
        "gadm41": {
          "global": {
            "url: "/region-metadata/gadm41/admin0" 
          },
          "country": {
            "url": "/region-metadata/gadm41/:level/:iso",  
            "level": [
              "admin1"
            ],
            "iso": [
              "AFG",
              "ALB"... etc 
            ]
          }
        }
      }
    }
  }
}
```

david-mears-2 commented:
> For the metadata response, does something like this seem better? It's more explicit about the valid urls and how the metadata provided fits into them. It does need to cover the issue of region metadata having two types of url, a "global" one and a "country" one (or suggest better names for these!)

Yes; slight preference to drop the top-level `datasets` key (making tile, regionMetadata top-level) since it's `gadm41` not `tile` that is the name of a dataset

EmmaLRussell commented:
Added a ticket for metadata changes: https://mrc-ide.myjetbrains.com/youtrack/issue/mrc-6831/Improve-grout-metadata-response

EmmaLRussell committed:
move region_metadata under top level data folder

EmmaLRussell committed:
rename region metadata controller methods

EmmaLRussell committed:
add note about region metadata endpoints to README

EmmaLRussell commented:
I think I've addressed all your comments, apart from those about rearranging the metadata response which will in this ticket: https://mrc-ide.myjetbrains.com/youtrack/issue/mrc-6831/Improve-grout-metadata-response - so should be ready for re-review.

david-mears-2 commented on tests/unit/controllers/regionMetadataController.spec.ts:
```diff
@@ -0,0 +1,86 @@
+import { describe, expect, test, vi, beforeEach } from "vitest";
+import { fs, vol } from "memfs";
+import * as jsonResponse from "../../../src/jsonResponse";
+import { jsonResponseSuccess } from "../../../src/jsonResponse";
```
My IDE points out that this import isn't used. Should this be caught by a formatter?

EmmaLRussell replied:
Yeah, I should switch on the "no-unused-vars" rule in eslint - this actually triggers quite a few errors, so I'll pop that in another PR.
